p2p/ping.go

```diff
@@ -0,0 +1,158 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/routing"
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+	ma "github.com/multiformats/go-multiaddr"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	pingInterval   = time.Second
+	pingProtectTag = "spacemesh-ping"
+)
+
+type pingStat struct {
+	numSuccess int
+	numFail    int
+}
+
+type Ping struct {
+	sync.Mutex
+	logger *zap.Logger
+	h      host.Host
+	peers  []peer.ID
+	pr     routing.PeerRouting
+	stats  map[peer.ID]*pingStat
+	cancel context.CancelFunc
+	eg     errgroup.Group
+}
+
+func NewPing(logger *zap.Logger, h host.Host, peers []peer.ID, pr routing.PeerRouting) *Ping {
+	p := &Ping{
+		logger: logger,
+		h:      h,
+		peers:  peers,
+		pr:     pr,
+		stats:  make(map[peer.ID]*pingStat),
+	}
+	return p
+}
+
+func (p *Ping) Start(ctx context.Context) {
+	if p.cancel != nil {
+		return
+	}
+	var startCtx context.Context
+	startCtx, p.cancel = context.WithCancel(context.Background())
+	for _, peer := range p.peers {
+		p.startForPeer(startCtx, peer)
+	}
+}
+
+func (p *Ping) Stop() {
+	if p.cancel == nil {
+		return
+	}
+	p.cancel()
+	p.cancel = nil
+	p.eg.Wait()
+}
+
+func (p *Ping) doPing(ctx context.Context, peerID peer.ID) (<-chan ping.Result, error) {
+	_, err := p.pr.FindPeer(ctx, peerID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find peer %s: %w", peerID, err)
+	}
+
+	return ping.Ping(ctx, p.h, peerID), nil
+}
+
+// runForPeer runs ping for the specific peer till an error occurs
+// or the context is canceled.
+func (p *Ping) runForPeer(ctx context.Context, peerID peer.ID, addrs []ma.Multiaddr) error {
+	pingCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	ch, err := p.doPing(pingCtx, peerID)
```

Comment:
I think you can just pass the parent context here:

```suggestion
	ch, err := p.doPing(ctx, peerID)
```

Reply:
If there's an error sent on the channel returned by libp2p's `Ping`, it appears that it stops working, most of the time, although this behavior is not documented and probably not intended, and may change in future versions. So what I want here is to always stop and restart ping when I get an error from the channel, thus I create a child context and cancel it when leaving `doPing`, which is then called again, after a delay, if there was a ping error.
I will add a comment to clarify that.